### DISCUSSION TITLE
Trying to work with a BNO085 IMU via I2C using Adafruit unified sensor library

### DISCUSSION THREAD

**Discussion by JakesWorkshop** [earlephilhower/arduino-pico#1324] [Q&A] [answered]
HI, I've recently been trying to port a program to this framework and it includes the Adafruit BNO08X library.
to start communications with the sensor a TwoWire needs to be created and passed to the initialization function the library provides. This fails for both Wire and Wire1, even if started correctly.
As far as I can tell, the initialization fails at creating a new Adafruit_I2CDevice using the specified TwoWire Bus.

Does anyone have any experience with the BNO085 on this Platform at all? Is there an alternative library to use?

I have previously got this to work using the I2C integration via MBedOS, where I created an MBedI2C TwoWire on the same pins as now and used it for the initialization and communication. This approach took a solid 7 ms to read data that was supposed to be available every 1 ms tho, which is why I would like to make this more efficient.

Edit: I also attempted to use the Library created by Sparkfun with no results. It is quite similar to the Adafruit library but slimmer. Use of those libraries might not actually be possible.

**Comment by mukyokyo** [comment, marked as answer]
I'm able to connect BNO085 to Pico's i2c and operate it using [ceva's drivers](https://github.com/ceva-dsp/sh2).

I had a little trouble treating the i2c reception, but solved the problem by reading the 4-byte header containing the data size and then reading all the data at that size.
```js
int IIC_Receive(uint8_t addr, uint8_t *dat, int buflen) {
  int result = 0;
  TSHTPHeader *h = (TSHTPHeader *)(void *)dat;
  int l = i2c_read_blocking(i2c0, addr, dat, 4, false);
    l = h->Length & ~0x8000;
  if (l > 0 && l <= buflen) {
    l = i2c_read_blocking(i2c0, addr, dat, l, false);
    if (l > 0) {
      l = h->Length & ~0x8000;
      if (l > 0) result = l;
    }
  }
  return result;
}
```
The data update cycle depends on the i2c bus speed and polling cycle, so requests that exceed this cycle are meaningless.
Incidentally, although not subject to Bno085 operation, it worked even when i2c speed was increased to 1MHz.

**Reply by JakesWorkshop** [reply]
Thank you kindly. 
I was not aware of the ceva driver yet but this and your method are just what I needed.

**Reply by jamescarruthers** [reply]
Any chance you could share your code and how you got this working correctly?

I'm having issues with all the other libraries and i2c.


> I'm able to connect BNO085 to Pico's i2c and operate it using [ceva's drivers](https://github.com/ceva-dsp/sh2).
> 
> 
> 
> I had a little trouble treating the i2c reception, but solved the problem by reading the 4-byte header containing the data size and then reading all the data at that size.
> 
> ```js
> 
> int IIC_Receive(uint8_t addr, uint8_t *dat, int buflen) {
> 
>   int result = 0;
> 
>   TSHTPHeader *h = (TSHTPHeader *)(void *)dat;
> 
>   int l = i2c_read_blocking(i2c0, addr, dat, 4, false);
> 
>     l = h->Length & ~0x8000;
> 
>   if (l > 0 && l <= buflen) {
> 
>     l = i2c_read_blocking(i2c0, addr, dat, l, false);
> 
>     if (l > 0) {
> 
>       l = h->Length & ~0x8000;
> 
>       if (l > 0) result = l;
> 
>     }
> 
>   }
> 
>   return result;
> 
> }
> 
> ```
> 
> The data update cycle depends on the i2c bus speed and polling cycle, so requests that exceed this cycle are meaningless.
> 
> Incidentally, although not subject to Bno085 operation, it worked even when i2c speed was increased to 1MHz.
> 
>

**Reply by mukyokyo** [reply]
An example of how it actually works can be found here.
[example](https://www.besttechnology.co.jp/modules/knowledge/?distributed%20edition%20PicoSHIELD#l7236c18)

**Reply by LazaroFilm** [reply]
I am using the SparkFun library which simplifies the i2c setup. The change I did is in the setup phase the library examples are checking if the BNO is detected, I replaced that by a while loop to keep on retrying because I noticed that the first few attempts would sometimes fail. You can add a counter to limit the number of attempts if you like. 
``` C
// Starting IMU

    while (myIMU.begin(BNO08X_ADDR, Wire, BNO08X_INT, BNO08X_RST) == false)
    {
        Serial.println("BNO08x not detected");
        delay(10);
    }
    Serial.println("BNO08x found!");
```

**Reply by jamescarruthers** [reply]
Thank you, this worked for me!

**Reply by jamescarruthers** [reply]
Thank you, this is super useful for my understanding.

**Comment by LazaroFilm** [comment]
I have succeeded in getting the Sparkfun library to work. What I did is instead of freezing when the BNO is not seen by Wire, just `delay(10)` and try again with a `while()` loop. It sometimes took a couple of attempts to get the connection. The other thing I read about is adding a 2K or 3K external pull up on top of the internal one. 

By the way, the SH2 library mentioned above is a dependency of Sparkfun's library.

**Comment by DineshSundrani7** [comment]
How to use BNO085 on 4B addr as it do not have D1 pin
